wallet/src/key_chain.rs

```diff
@@ -17,184 +17,619 @@
+use common::chain::config::BIP44_PATH;
 use common::chain::ChainConfig;
 use crypto::key::extended::{ExtendedKeyKind, ExtendedPrivateKey, ExtendedPublicKey};
+use crypto::key::hdkd::child_number::ChildNumber;
 use crypto::key::hdkd::derivable::{Derivable, DerivationError};
 use crypto::key::hdkd::derivation_path::DerivationPath;
-use std::str::FromStr;
+use crypto::key::hdkd::u31::U31;
+use serialization::{Decode, Encode};
+use std::collections::BTreeMap;
+use std::convert::TryInto;
 use std::sync::Arc;
 use storage::Backend;
-use wallet_storage::{Store, WalletStorageRead};
+use wallet_storage::{StoreTxRo, StoreTxRw, WalletStorageRead, WalletStorageWrite};
+use wallet_types::account_info::KeychainUsageState;
+use wallet_types::{
+    AccountAddressId, AccountId, AccountInfo, DeterministicAccountInfo, KeyContent, KeyId,
+    KeyIdPrefix,
+};
 use zeroize::Zeroize;
 
-const KEY_KIND: ExtendedKeyKind = ExtendedKeyKind::Secp256k1Schnorr;
+/// The number of nodes in a BIP44 path
+const BIP44_PATH_LENGTH: usize = 5;
+/// The index of key_purpose
+const BIP44_KEY_PURPOSE_INDEX: usize = 3;
+/// The index of the usable key in the BIP44 hierarchy
+const BIP44_KEY_INDEX: usize = 4;
+/// Default cryptography type
+const DEFAULT_KEY_KIND: ExtendedKeyKind = ExtendedKeyKind::Secp256k1Schnorr;
+/// Default size of the number of unused addresses that need to be checked after the
+/// last used address.
 const LOOKAHEAD_SIZE: u16 = 100;
 
 /// KeyChain errors
 #[derive(thiserror::Error, Debug, Eq, PartialEq)]
 pub enum KeyChainError {
+    #[error("Wallet database error: {0}")]
+    DatabaseError(#[from] wallet_storage::Error),
     #[error("Bip39 error: {0}")]
     Bip39(bip39::Error),
     #[error("Key derivation error: {0}")]
     Derivation(#[from] DerivationError),
     #[error("Address error: {0}")]
     Address(#[from] AddressError),
+    #[error("Key chain is locked")]
+    KeyChainIsLocked,
+    #[error("No account found")] // TODO implement display for AccountId
+    NoAccountFound(AccountId),
+    #[error("Invalid BIP44 derivation path format: {0}")]
+    InvalidBip44DerivationPath(DerivationPath),
+    #[error("Could not load key chain")]
+    CouldNotLoadKeyChain,
+    #[error("The provided keys do not belong to the same hierarchy")]
+    KeysNotInSameHierarchy,
+    #[error("Invalid key purpose index {0}")]
+    InvalidKeyPurpose(ChildNumber),
+    #[error("Only one root key is supported")]
+    OnlyOneRootKeyIsSupported,
+    #[error("Cannot issue more keys, lookahead exceeded")]
+    LookAheadExceeded,
 }
 
 /// Result type used for the key chain
 type KeyChainResult<T> = Result<T, KeyChainError>;
 
 /// The usage purpose of a key i.e. if it is for receiving funds or for change
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Encode, Decode)]
 pub enum KeyPurpose {
     /// This is for addresses created for receiving funds that are given to the user
-    ReceiveFunds = 0,
+    #[codec(index = 0)]
+    ReceiveFunds,
```

Comment:
Why did you make this change? The derive macro for `Encode` and `Decode` recognises the `EnumArm = X` syntax and uses that for encoding and decoding.